LNK-4284: Expose consistent API Info

seanmcilvenna opened the pull request:
### 🛠️ Description of Changes

All services now have /api/XXX/info endpoint that return consistent information including:

* version
* productVersion
* serviceName
* build
* commit

Removed productVersion from all existing `appSettings.json` files so that those could be written at build time, instead.

`version` now automatically represents the assembly version, but may still be overridden by config if we decide to do so later. In java services, `version` represents the `BuildProperties` information that is set by Maven during the `build-info` execution goal.

### 🧪 Testing Performed

* Built and ran the entire stack in Docker Compose and tested in service's /api/XXX/info endpoint
* Made changes to environment variables to test that configurable properties likee serviceName and productVersion can be overridden

### 🧑‍🔬 Unit Testing

Since there is no actual business logic in this change, I did not write tests. All of these changes represent run-time configurations.

### 📓 Documentation Updated

Config documentation has been updated to indicate the properties that can be set at runtime, and what properties are automatically set (such as `version`).


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added a unified service info endpoint (/api/{service}/info) and health endpoints across services; Java endpoints added for measureeval/validation.

- **Refactor**
  - Standardized service metadata handling and startup configuration; Swagger now displays configured service names and includes health metadata.

- **Documentation**
  - New .NET/Java docs describing the service info and health endpoints and configuration examples.

- **Chores**
  - Updated docker-compose/env keys to ServiceName/ProductVersion and refreshed Swagger/OpenAPI specs and tooling.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

seanmcilvenna committed:
Alignment/consistency across .net services to expose /api/<service>/info in the same way with the correct information

seanmcilvenna committed:
Adding service information config model and a controller to each of the java services to return the service information config.

seanmcilvenna committed:
Adding documentation about service information configuration

arianamihailescu commented on Java/shared/src/main/java/com/lantanagroup/link/shared/config/ServiceInformationConfig.java:
```diff
@@ -0,0 +1,30 @@
+    private String serviceName;
+    private String version;
+    private String productVersion;
+    private String build;
```
Do we need the commit field as well in the ServiceInformationConfig class? 
And if we add the commit l can we use the git-commit-id-plugin to get it from git like CodeRabbitai suggests? Will that make the script that injects the commit unnecessary (Scripts/git_commit.ps1)?

seanmcilvenna replied:
We will already need to update the build pipelines to set the commit hash for .net services... So, for consistency, I'm not going to try to automate this in the java services; we'll just have the build pipeline do this for all services by (probably) updating the `appSettings.json` and `application.yml` files to specify `ServiceInformation__Commit` and `service-information.commit` prior to building so that it's baked into the DLLs and JARs.

seanmcilvenna commented on docker-compose.yml:
```diff
@@ -735,6 +737,8 @@
     image: link-measureeval
     container_name: link-measureeval
     environment:
+      service-information.service-name: "Link Measure Evaluation Service"
+      service-information.product-version: dev
```
but there are lots of examples in this same service where spring _does_ resolve env vars with dots/hyphens. not changing because it does seem to work, contrary to the suggestion.

seanmcilvenna commented on docs/docs/config/dotnet.mdx:
````diff
@@ -116,6 +116,30 @@ Not all services use every URL. The configuration is provided for completeness a
 | ServiceRegistry__ReportServiceUrl          | Base URL for the Report service           |
 | ServiceRegistry__SubmissionServiceUrl      | Base URL for the Submission service       |
 
+### Example Configuration in JSON
+
+```json
+{
+  "ServiceRegistry": {
+    "AccountServiceUrl": "https://localhost:7001",
+    "AuditServiceUrl": "https://localhost:7334",
+    "CensusServiceUrl": "https://localhost:7234",
+    "DataAcquisitionServiceUrl": "https://localhost:7194",
+    "MeasureServiceUrl": "https://localhost:7135",
+    "NormalizationServiceUrl": "https://localhost:7038",
+    "NotificationServiceUrl": "https://localhost:7434",
+    "QueryDispatchServiceUrl": "https://localhost:7534",
+    "ReportServiceUrl": "https://localhost:7110",
+    "SubmissionServiceUrl": "https://localhost:7046",
+    "TenantService": {
+      "TenantServiceUrl": "https://localhost:7332",
+      "CheckIfTenantExists": true,
+      "GetTenantRelativeEndpoint": "facility/"
+    }
+  }
+}
+```
````
I'm not sure what table you're referring. The table in the docs/docs/config/dotnet.mdx shows `TenantService` _is_ nested below `ServiceRegistry`. Also confirmed that the object model in LantanaGroup.Link.Shared.Application.Models.Configs.ServiceRegistry has a sub-property called `TenantService` that is of type `TenantServiceRegistration`. So, that object model seems to align with the tenant service's current appSettings.json file.

seanmcilvenna commented on docs/docs/config/dotnet.mdx:
````diff
@@ -139,28 +163,18 @@
-    }
-  }
-}
-```
+The ServiceInformation configuration allows exposing build, version and deployment information through the `/api/XXX/info` (i.e. `/api/audit/info`) endpoint of each service. This information is automatically populated from assembly metadata and git information during build time.
+
+| Property                           | Description                         | Required | Default Value                   | Secret? |
+|------------------------------------|-------------------------------------|----------|---------------------------------|---------|
+| ServiceInformation__ServiceName    | Display name of the service         | No       | Varies service by service       | No      |
+| ServiceInformation__Version        | Version of the service              | No       | Assembly ver.                   | No      |
+| ServiceInformation__ProductVersion | Version of the product as a whole   | No       | N/A                             | No      |
+| ServiceInformation__Build          | Build number from CI/CD pipeline    | No       | "dev" when using Docker Compose | No      |
+| ServiceInformation__Commit         | Git commit hash of deployed version | No       | None                            | No      |
+
+The service automatically extracts information from .NET assembly version in the AssemblyInfo.cs file or project file
````
Ignoring nitpick

seanmcilvenna commented on Java/shared/src/main/java/com/lantanagroup/link/shared/config/ServiceInformationConfig.java:
```diff
@@ -0,0 +1,30 @@
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.boot.info.BuildProperties;
+
+@ConfigurationProperties("service-information")
+@Getter
+@Setter
+public class ServiceInformationConfig {
+    private String serviceName;
+    private String version;
+    private String productVersion;
+    private String build;
+
+    @Autowired
+    @JsonIgnore
+    private BuildProperties buildProperties;
+
+    @PostConstruct
+    public void initializeVersion() {
+        if (this.version == null || this.version.isEmpty()) {
+            this.version = buildProperties != null ? buildProperties.getVersion() : "UNKNOWN";
+        }
+    }
+}
```
The .net services do not have an equivalent to this `GitProperties` concept and the build pipelines will already have to be responsible for configuring/setting the git commit hash. For consistency, I'm not going to try to automate that in code for java services.

seanmcilvenna committed:
Fixing docker compose to have correct `ServiceName` env properties
Updating ApiInfoController to explicitly set GetMapping instead of a generic RequestMapping that defaults to GET

seanmcilvenna committed:
Per coderabbitai suggestion, updating swagger docs to include /info endpoints
While doing that, discovered that /health endpoints weren't included in swagger docs and created a document filter to add the health endpoints dynamically
Fixed download script to update docs with the latest swagger specs

seanmcilvenna committed:
Updates to documentation for swagger specs using download-swagger-specs.ps1 script

seanmcilvenna commented on Java/measureeval/src/main/java/com/lantanagroup/link/measureeval/controllers/ApiInfoController.java:
```diff
@@ -0,0 +1,20 @@
+        this.serviceInformationConfig = serviceInformationConfig;
+    }
+
+    @GetMapping("/info")
+    public ServiceInformationConfig info() {
+        return this.serviceInformationConfig;
+    }
```
Ignoring nitpick

seanmcilvenna committed:
Making autowire optional for BuildProperties so-as not to affect local builds/runs in IDEs

seanmcilvenna commented on docs/domains/Report/services/NormalizationService/openapi.yml:
```diff
@@ -1,28 +1,144 @@
-﻿openapi: 3.0.1
+﻿openapi: 3.0.4
 info:
-  version: 0.2.0
+  version: 0.3.0
   title: Normalization
```
ignoring nitpick

seanmcilvenna commented on docs/domains/Report/services/NormalizationService/openapi.yml:
```diff
@@ -1,28 +1,144 @@
-  /api/Normalization:
-    post:
+  /api/normalization/info:
+    get:
       tags:
       - Normalization
-      summary: Save a config for a facility.
+      responses:
+        "200":
+          description: OK
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ServiceInformation'
```
ignoring nitpick

seanmcilvenna commented on docs/domains/Report/services/NormalizationService/openapi.yml:
```diff
@@ -1,28 +1,144 @@
+        schema:
+          $ref: '#/components/schemas/SortOrder'
+      - name: pageSize
+        in: query
+        schema:
+          type: integer
+          format: int32
+          default: 10
+      - name: pageNumber
+        in: query
+        schema:
+          type: integer
+          format: int32
+          default: 1
```
ignoring nitpick

seanmcilvenna commented on Scripts/download-swagger-specs.ps1:
```diff
@@ -0,0 +1,56 @@
+        # Download Swagger JSON
+        $jsonContent = Invoke-RestMethod -Uri $url
+        
+        $jsonContent.info = @{
+            title   = $jsonContent.info.title
+            version = $NewVersion
+        }
+        
```
ignoring. the script works the way I want it to for this.

seanmcilvenna commented on Scripts/download-swagger-specs.ps1:
```diff
@@ -0,0 +1,56 @@
+            version = $NewVersion
+        }
+        
+        $jsonContent.PSObject.Properties.Remove("servers")
```
ignoring nitpick

seanmcilvenna commented on Scripts/download-swagger-specs.ps1:
```diff
@@ -0,0 +1,56 @@
+        }
+
+        # Write YAML to file
+        Set-Content -Path $outputPath -Value $yamlContent -Encoding UTF8
```
ignoring nitpick

seanmcilvenna commented on Scripts/download-swagger-specs.ps1:
```diff
@@ -0,0 +1,56 @@
+
+        Write-Host "Saved YAML to $outputPath"
+    } catch {
+        Write-Error "Failed to process $url due to $_"
+    }
```
ignoring nitpick

seanmcilvenna commented on DotNet/Tenant/Program.cs:
```diff
@@ -267,11 +268,12 @@ static void SetupMiddleware(WebApplication app)
             app.MapControllers();
 
-            //map health check middleware
+            //map health check middleware and info endpoint
             app.MapHealthChecks("/health", new HealthCheckOptions
             {
                 ResponseWriter = UIResponseWriter.WriteHealthCheckUIResponse
             });
+            app.MapInfo(Assembly.GetExecutingAssembly(), app.Configuration, "facility");
 
```
I want /info endpoints to be anonymously accessible

seanmcilvenna commented on DotNet/Shared/Application/Health/HealthChecksFilter.cs:
```diff
@@ -0,0 +1,112 @@
+        };
+
+        // Get registered health checks
+        var healthCheckService = _serviceProvider.GetRequiredService<HealthCheckService>();
+        var registrations = _serviceProvider.GetRequiredService<IOptions<HealthCheckServiceOptions>>().Value.Registrations;
```
ignoring nitpick

seanmcilvenna commented on docs/domains/Report/services/NormalizationService/openapi.yml:
```diff
@@ -1,28 +1,144 @@
+        schema:
+          type: string
+          format: uuid
+      - name: sortBy
+        in: query
+        schema:
+          type: string
+          default: CreateDate
+      - name: sortOrder
+        in: query
+        schema:
+          $ref: '#/components/schemas/SortOrder'
```
ignoring nitpick

arianamihailescu reviewed (approved):
Looks  good
